sql/src/main/java/io/crate/execution/dml/upsert/ShardUpsertRequest.java

```diff
@@ -253,16 +255,59 @@ public int hashCode() {
         @Nullable
         private Object[] insertValues;
 
+        /**
+         * The URI the item was read from.
+         * It is only needed ony the sender to merge the write response into the related result object.
+         * DO NOT STREAM!
+         */
+        @Nullable
+        private BytesRef sourceUri;
+
+        /**
+         * A possible failure while reading the item.
+         * It is only needed ony the sender to merge the write response into the related result object.
+         * DO NOT STREAM!
+         */
+        @Nullable
+        private String readFailure;
+
+        /**
+         * A possible failure processing the source URI.
+         * It is only needed ony the sender to merge the write response into the related result object.
+         * DO NOT STREAM!
+         */
+        @Nullable
+        private String sourceUriFailure;
+
         public Item(String id,
                     @Nullable Symbol[] updateAssignments,
                     @Nullable Object[] insertValues,
                     @Nullable Long version) {
+            this(id, updateAssignments, insertValues, version, null);
+        }
+
+        public Item(String readFailure, @Nullable BytesRef sourceUri) {
+            this("failed", null, null, null, sourceUri);
```

Comment:
yes because failed items are just marker items to generate error summary from. any suggestion to express it better that it is safe in that case?

Reply:
Not sure, we could use `""`, or `null` but that requires a bunch of other checks. 

In general, would it be nicer to have a dedicated subclass for Item failures? That way we wouldn't have to add to the normal Item class.

Reply:
I think, empty or null values would be even worse as there won't be any hint what happens (error item or something wrong?).
Introducing a subclass would require some detection (instance of, enum, ...) which then must happen on every item which I'd like to avoid.